metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -129,155 +161,519 @@ func getResource(resourceName string) kubernetes.Resource {
+	options kubernetes.WatchOptions,
+	client k8sclient.Interface,
+	resourceWatchers *Watchers,
+	namespace string,
+	extraWatcher bool) (bool, error) {
+
+	// We need to check the node scope to decide on whether a watcher should be updated or not.
+	nodeScope := false
+	if options.Node != "" {
+		nodeScope = true
+	}
+	// The nodescope for extra watchers node, namespace, replicaset and job should be always false.
+	if extraWatcher {
+		nodeScope = false
+		options.Node = ""
+	}
+
+	resourceWatchers.lock.Lock()
+	defer resourceWatchers.lock.Unlock()
+
+	// Check if a watcher for the specific resource already exists.
+	resourceMetaWatcher, ok := resourceWatchers.metaWatchersMap[resourceName]
+
+	// If it does not exist, create the resourceMetaWatcher.
+	if !ok {
+		// Check if we need to add namespace to the watcher's options.
+		if isNamespaced(resourceName) {
+			options.Namespace = namespace
+		}
+		watcher, err := kubernetes.NewNamedWatcher(resourceName, client, resource, options, nil)
+		if err != nil {
+			return false, err
+		}
+		resourceWatchers.metaWatchersMap[resourceName] = &metaWatcher{
+			watcher:         watcher,
+			started:         false, // not started yet
+			metadataObjects: make(map[string]bool),
+			enrichers:       make(map[string]*enricher),
+			metricsetsUsing: make([]string, 0),
+			restartWatcher:  nil,
+			nodeScope:       nodeScope,
+		}
+		return true, nil
+	} else if resourceMetaWatcher.nodeScope != nodeScope && resourceMetaWatcher.nodeScope {
+		// It might happen that the watcher already exists, but is only being used to monitor the resources
+		// of a single node(e.g. created by pod metricset). In that case, we need to check if we are trying to create a new watcher that will track
+		// the resources of whole cluster(e.g. in case of state_pod metricset).
+		// If it is the case, then we need to update the watcher by changing its watch options (removing options.Node)
+		// A running watcher cannot be updated directly. Instead, we must create a new one with the correct watch options.
+		// The new restartWatcher must be identical to the old watcher, including the same handler function, with the only difference being the watch options.
+
+		if isNamespaced(resourceName) {
+			options.Namespace = namespace
+		}
+		restartWatcher, err := kubernetes.NewNamedWatcher(resourceName, client, resource, options, nil)
+		if err != nil {
+			return false, err
+		}
+		// update the handler of the restartWatcher to match the current watcher's handler.
+		restartWatcher.AddEventHandler(resourceMetaWatcher.watcher.GetEventHandler())
+		resourceMetaWatcher.restartWatcher = restartWatcher
+		resourceMetaWatcher.nodeScope = nodeScope
+	}
+	return false, nil
+}
+
+// addToMetricsetsUsing adds metricset identified by metricsetUsing to the list of resources using the shared watcher
+// identified by resourceName.
+func addToMetricsetsUsing(resourceName string, metricsetUsing string, resourceWatchers *Watchers) {
```

Comment:
`removeFromMetricsetsUsing` provides information about the lock in the docstring, please do the same here. Add that the caller should not be holding the lock.

Reply:
What about changing this function to not hold the lock and having the caller handle it? That way its consistent with `removeFromMetricsetsUsing`?

Reply:
Done